### PR TITLE
OMWAPPI-1620: org.rdk.Wifi implementation

### PR DESCRIPTION
**Pull request by tomasz-karczewski-red**
Lgi specific implementation based on dbus ifaces:

com.lgi.rdk.utils.networkconfig1
com.lgi.rdk.utils.wifimanagement1

Only following methods and events are supported
for now:

methods:
setSignalThresholdChangeEnabled
getCurrentState
getConnectedSSID

events:
onWIFIStateChanged
onWifiSignalThresholdChanged
onSSIDsChanged

(this should be enough for Amazon app)

### CODE REVIEW

File: WifiManager/impl_lg/WifiManagerState.cpp [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+      LOGWARN("failed to fetch interfaces via networkconfig1_GetInterfaces");
+   }
+   return "";
+}
+
+const std::string &WifiManagerState::getWifiInterfaceName()
+{
+   static std::mutex interface_name_mutex;
+   std::lock_guard<std::mutex> lock(interface_name_mutex);
+   static std::string name = WifiManagerState::fetchWifiInterfaceName();
```
Comment (dawid-trendota-red) [inline WifiManager/impl_lg/WifiManagerState.cpp:233]: It is a static value, which is protected only within this function (and it is used outside this function)

File: WifiManager/impl_lg/WifiManagerSignalThreshold.cpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+using namespace WPEFramework::Plugin;
+
+namespace {
+
+    const float signalStrengthThresholdExcellent = -50.0f;
+    const float signalStrengthThresholdGood = -60.0f;
+    const float signalStrengthThresholdFair = -67.0f;
+
+    // compare onemw-src/networking/om-netconfig/src/wpa/WifiUtils.cpp, WifiUtils::decibels2quality(int aDB), where lQuality = 2 * (aDB + 100), for aDB in (-100,-50)
+    std::string quality2decibels(std::string qualityStr) {
```
[the remark this answers is not in the document]
Comment (tomasz-karczewski-red) [reply WifiManager/impl_lg/WifiManagerSignalThreshold.cpp:36]: ok

File: WifiManager/impl_lg/WifiManagerSignalThreshold.cpp [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        return "";
+    }
+
+    void getSignalData(float &signalStrengthOut, std::string &strengthOut) {
+        JsonObject response;
+
+        string signalStrength = retrieveValue();
+
+        signalStrengthOut = 0.0f;
+        if (!signalStrength.empty())
```
[the remark this answers is not in the document]
Comment (tomasz-karczewski-red) [reply WifiManager/impl_lg/WifiManagerSignalThreshold.cpp:75]: this is copy pasted from non-lgi impl, so I'd leave it as it is for now

File: WifiManager/impl_lg/WifiManagerSignalThreshold.cpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+    uint32_t result = WifiManager::getInstance().getCurrentState(parameters, response);
+    if (result != 0)
+    {
+        LOGINFO("wifiManager.getCurrentState result = %d", result);
+        return;
+    }
+    else if (response.HasLabel("state"))
+    {
+        int64_t number = std::stoi(response["state"].String());
```
[the remark this answers is not in the document]
Comment (tomasz-karczewski-red) [reply WifiManager/impl_lg/WifiManagerSignalThreshold.cpp:161]: this is copy pasted from non-lgi impl, so I'd leave it as it is for now

File: WifiManager/impl_lg/WifiManagerSignalThreshold.cpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        state = (WifiState) number;
```
[the remark this answers is not in the document]
Comment (tomasz-karczewski-red) [reply WifiManager/impl_lg/WifiManagerSignalThreshold.cpp:162]: this is copy pasted from non-lgi impl, so I'd leave it as it is for now

File: WifiManager/impl_lg/WifiManagerSignalThreshold.cpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    }
+    else
+    {
+        LOGINFO("no state attribute");
+        return;
+    }
+
+    if(changeEnabled)
+    {
+        if (state == WifiState::CONNECTED)
```
[the remark this answers is not in the document]
Comment (tomasz-karczewski-red) [reply WifiManager/impl_lg/WifiManagerSignalThreshold.cpp:173]: this is copy pasted from non-lgi impl, so I'd leave it as it is for now

File: WifiManager/impl_lg/WifiManagerState.cpp [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+      LOGWARN("No 'wifi' interface found");
+      // TODO: throw an exception?
+   }
+   else
+   {
+      // register for status updates
+      dbus.registerStatusChanged(std::bind(&WifiManagerState::statusChanged, this, std::placeholders::_1, std::placeholders::_2));
+      // get current wifi status
+      InterfaceStatus status;
+      const std::string iname = getWifiInterfaceName();
```
[the remark this answers is not in the document]
Comment (tomasz-karczewski-red) [reply WifiManager/impl_lg/WifiManagerState.cpp:45]: ok

File: WifiManager/impl_lg/WifiManagerState.cpp (hunk 1 above)
Comment (tomasz-karczewski-red) [reply WifiManager/impl_lg/WifiManagerState.cpp:233]: on another look, no synchro is needed here (local static initialization is thread-safe per standard, and we later return the const &)

File: WifiManager/impl_lg/dbus/DBusClient.cpp [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+                }
+            }
+        }
+        else
+        {
+            LOGERR("Failed to create glib main loop");
+            m_dbus_data_future.set_value(nullptr);
+        }
+
+        if (error) g_error_free(error);
```
[the remark this answers is not in the document]
Comment (tomasz-karczewski-red) [reply WifiManager/impl_lg/dbus/DBusClient.cpp:373]: ok

File: WifiManager/impl_lg/WifiManagerSignalThreshold.cpp [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        thread.join();
+    }
+    running = false ;
+}
+
+
+void WifiManagerSignalThreshold::setSignalThresholdChangeEnabled(bool enable)
+{
+    LOGINFO("setSignalThresholdChangeEnabled: enable %s", enable ? "true":"false");
+    running = enable;
```
[the remark this answers is not in the document]
Comment (tomasz-karczewski-red) [reply WifiManager/impl_lg/WifiManagerSignalThreshold.cpp:223]: no really, it seems :) i'll make it atomic

**Review by dawid-trendota-red — APPROVED** [review]

**Pull request merged**